File: packages/guards/src/pii-es.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type PiiEsEntity = 'dni' | 'nie' | 'cif' | 'phone-es' | 'iban-es';
interface PiiEsOptions { entities: PiiEsEntity[]; action: 'block' | 'warn' | 'mask'; }
interface PiiEsMatch { type: PiiEsEntity; value: string; start: number; end: number; }

const PATTERNS: Record<PiiEsEntity, RegExp> = {
  dni: /\b\d{8}[A-Z]\b/g,
  nie: /\b[XYZ]\d{7}[A-Z]\b/g,
  cif: /\b[ABCDEFGHJNPQRSUVW]\d{7}[A-J0-9]\b/g,
  'phone-es': /\b(?:\+?34[-.\s]?)?[6-9]\d{2}[-.\s]?\d{3}[-.\s]?\d{3}\b/g,
  'iban-es': /\bES\d{2}\s?\d{4}\s?\d{4}\s?\d{2}\s?\d{10}\b/g,
};
const MASK: Record<PiiEsEntity, string> = { dni: '[DNI]', nie: '[NIE]', cif: '[CIF]', 'phone-es': '[Telefono]', 'iban-es': '[IBAN]' };

function detect(text: string, entities: PiiEsEntity[]): PiiEsMatch[] {
  const m: PiiEsMatch[] = [];
  for (const e of entities) { const re = new RegExp(PATTERNS[e].source, 'g'); let x; while ((x = re.exec(text))) m.push({ type: e, value: x[0], start: x.index, end: x.index + x[0].length }); }
  return m.sort((a, b) => b.start - a.start);
}

export function piiEs(options: PiiEsOptions): Guard {
  return { name: 'pii-es', version: '0.1.0', description: 'Spanish PII (DNI, NIE, CIF, phone, IBAN)', category: 'privacy', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const matches = detect(text, options.entities); const triggered = matches.length > 0;
      if (!triggered) return { guardName: 'pii-es', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      if (options.action === 'mask') { let r = text; for (const m of matches) r = r.slice(0, m.start) + MASK[m.type] + r.slice(m.end); return { guardName: 'pii-es', passed: true, action: 'override', overrideText: r, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type }) => ({ type })) } }; }
      return { guardName: 'pii-es', passed: false, action: options.action, message: `Spanish PII: ${[...new Set(matches.map(m => m.type))].join(', ')}`, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
    },
  };
}

File: python/open_guardrail/guards/pii_es.py
"""Spanish PII detection and masking guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "dni": re.compile(r"\b\d{8}[A-Z]\b"),
    "nie": re.compile(r"\b[XYZ]\d{7}[A-Z]\b"),
    "nss": re.compile(r"\b\d{2}/?\d{8}/?\d{2}\b"),
    "iban-es": re.compile(r"\bES\d{2}\s?\d{4}\s?\d{4}\s?\d{2}\s?\d{10}\b", re.I),
    "phone-es": re.compile(r"\b(?:\+34|0034)\s?[6-9]\d{2}[\s\-]?\d{3}[\s\-]?\d{3}\b"),
    "cif": re.compile(r"\b[ABCDEFGHJNPQRSUVW]\d{7}[A-Z0-9]\b"),
}

_MASK_LABELS: dict[str, str] = {
    "dni": "[DNI]",
    "nie": "[NIE]",
    "nss": "[NSS]",
    "iban-es": "[IBAN-ES]",
    "phone-es": "[TELÉFONO-ES]",
    "cif": "[CIF]",
}


def _detect(text: str, entities: list[str]) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append({"type": entity, "value": m.group(), "start": m.start(), "end": m.end()})
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        result = result[: m["start"]] + _MASK_LABELS.get(m["type"], "[PII]") + result[m["end"] :]
    return result


class _PiiEs:
    def __init__(self, *, action: str = "block", entities: Optional[List[str]] = None) -> None:
        self.name = "pii-es"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000
        if not matches:
            return GuardResult(guard_name="pii-es", passed=True, action="allow", latency_ms=round(elapsed, 2))
        detected = [{"type": m["type"], "value": m["value"]} for m in matches]
        if self.action == "mask":
            return GuardResult(guard_name="pii-es", passed=True, action="override", override_text=_mask(text, matches), latency_ms=round(elapsed, 2), details={"detected": detected})
        return GuardResult(guard_name="pii-es", passed=False, action=self.action, message="Spanish PII detected", latency_ms=round(elapsed, 2), details={"detected": detected})


def pii_es(*, action: str = "block", entities: Optional[List[str]] = None) -> _PiiEs:
    return _PiiEs(action=action, entities=entities)
